mailbox/noise.go

```diff
@@ -787,27 +850,52 @@ func (b *Machine) RecvActTwo(actTwo [ActTwoSize]byte) error {
 	}
 	b.mixKey(es)
 
-	// The payload sent during this second act by the responder is to be
-	// interpreted as an additional piece of authentication information.
-	payload, err := b.DecryptAndHash(p[:])
-	if err != nil {
+	if b.handshakeVersion == HandshakeVersion0 {
+		var p [ActTwoPayloadSizeV0 + 16]byte
+		if _, err := io.ReadFull(r, p[:]); err != nil {
+			return err
+		}
+
+		// The payload sent during this second act by the responder is
+		// to be interpreted as an additional piece of authentication
+		// information.
+		payload, err := b.DecryptAndHash(p[:])
+		if err != nil {
+			return err
+		}
+
+		// If the payload is a non-zero length, then we'll assume it's
+		// the auth data and attempt to fully decode it.
+		if len(payload) == 0 {
+			return err
+		}
+
+		payloadLen := binary.BigEndian.Uint16(payload[:2])
+		b.authData = make([]byte, payloadLen)
+
+		payloadReader := bytes.NewReader(payload[2:])
+		_, err = payloadReader.Read(b.authData)
```

Comment:
See comment above, thsi needs to be handled distinctly (explicit handshake version handling).